support-services/src/main/scala/com/gu/support/redemption/generator/GiftGenerator.scala

```diff
@@ -0,0 +1,84 @@
+package com.gu.support.redemption.generator
+
+import java.security.SecureRandom
+
+import com.gu.support.redemption.generator.ConstructCode.GenerateGiftCode
+
+object GiftGenerator {
+
+  lazy val randomGiftCodes = {
+    val gen = new SecureRandom()
+    val ints = Iterator.continually(gen.nextInt())
+    apply(ints)
+  }
+
+  def apply(random: Iterator[Int]): Iterator[GenerateGiftCode] =
+    IntsToTypableString(random).map(ConstructCode.apply)
+
+}
+
+sealed abstract class GiftDuration(val code: String)
+
+object GiftDuration {
+
+  case object Gift3Month extends GiftDuration("03")
+
+  case object Gift6Month extends GiftDuration("06")
+
+  case object Gift12Month extends GiftDuration("12")
+
+}
+
+object ConstructCode {
+
+  private val prefix = "gd"
+
+  case class GiftCode private(value: String) extends AnyVal
+
+  object GiftCode {
+    def apply(value: String): Option[GiftCode] =
+      Some(value)
+        .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{6}"""))
+        .map(new GiftCode(_))
+  }
+
+  trait GenerateGiftCode {
+    def apply(duration: GiftDuration): GiftCode
+  }
+
+  def apply(code: IntsToTypableString.Code): GenerateGiftCode =
+    (duration: GiftDuration) => {
+      val init = prefix + duration.code + "-"
+      GiftCode(init + code.value).get
+    }
+
+}
+
+object IntsToTypableString {
```

Comment:
I'm not that keen on `Typable` could we call this `IntsToCode` or `IntsToCodeSuffix` or I think my preference would be `CodeSuffixGenerator`?